libc/stdio_file.c

```diff
@@ -0,0 +1,785 @@
+/*
+ * Phoenix-RTOS
+ *
+ * libc-tests
+ *
+ * libphoenix functions tests from stdio/file.c
+ * 
+ * TESTED:
+ * fopen, fclose, fdopen, freopen,
+ * fwrite, fread,
+ * putc, fputc, fputs,
+ * getc, fgetc, fgets,
+ * ungetc,
+ * getline,
+ * fseek, fseeko, rewind,
+ * ftell,
+ * fileno, feof, remove,
+ * ferror, clearerr,
+ * setvbuf, setbuf, fflush,
+ * 
+ * UNTESTED:
+ * puts, gets < needs writing to stdin/unimplemented
+ * popen, pclose, tmpfile < not usable on all targets
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Mateusz Niewiadomski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <stdio.h>
+#include <errno.h>
+#include <string.h>
+#include <limits.h>
+#include <stdlib.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include <unity_fixture.h>
+
+
+FILE *fd, *fd2;
+char stdpath[] = "stdio_file_test";
+char c, buf[20];
+int filedes;
+char toolongpath[PATH_MAX + 16];
+
+/* 
+Tets group for:
+fopen, fclose,
+fdopen, freopen
+*/
+TEST_GROUP(fopenfclose);
+
+TEST_SETUP(fopenfclose)
+{
+	memset(toolongpath, 'a', PATH_MAX + 15);
+	toolongpath[PATH_MAX + 15] = '\0';
+}
+
+
+TEST_TEAR_DOWN(fopenfclose)
+{ /* empty */
+}
+
+
+int checkdescriptor(FILE *fd)
+{
+	return (fd == NULL) ? 0 : !fclose(fd);
+}
+
+
+void assert_fopen_error(char *path, char *opts, int errnocode)
+{
+	fd = fopen(path, opts);
+	TEST_ASSERT_FALSE(checkdescriptor(fd));
+	TEST_ASSERT_EQUAL_INT(errnocode, errno);
+}
+
+
+void assert_fopen_success(char *path, char *opts)
+{
+	fd = fopen(path, opts);
+	TEST_ASSERT_TRUE(checkdescriptor(fd));
+}
+
+
+TEST(fopenfclose, fopenfclose_file)
+{
+	/* not existing file opening without creating */
+	assert_fopen_error(stdpath, "r", ENOENT);
+	assert_fopen_error(stdpath, "r+", ENOENT);
+	/* opening file with creation */
+	assert_fopen_success(stdpath, "w");
+	assert_fopen_success(stdpath, "a");
+	assert_fopen_success(stdpath, "w+");
+	assert_fopen_success(stdpath, "a+");
+	/* opening existing file for read */
+	assert_fopen_success(stdpath, "r");
+	assert_fopen_success(stdpath, "r+");
+}
+
+
+TEST(fopenfclose, fopenfclose_opendir)
+{
+	/* open directory */
+	assert_fopen_success("/dev/", "r");
+	assert_fopen_error("/dev/", "w", EISDIR);
+}
+
+TEST(fopenfclose, fopenfclose_zeropath)
+{
+	/* open null or zero path */
+	assert_fopen_error("", "r", ENOENT);
+	assert_fopen_error(NULL, "r", EINVAL);
+	assert_fopen_error("", "w", ENOENT);
+	assert_fopen_error(NULL, "w", EINVAL);
+}
+
+TEST(fopenfclose, fopenfclose_wrongflags)
+{
+	/* open with no flags/wrong flags/null flags */
+	assert_fopen_error(stdpath, "", EINVAL);
+	assert_fopen_error(stdpath, "phoenix-rtos", EINVAL);
+	assert_fopen_error(stdpath, NULL, EINVAL);
+}
+
+TEST(fopenfclose, fopenfclose_toolongname)
+{
+	/* open file with too long name */
+	assert_fopen_error(toolongpath, "w", ENAMETOOLONG);
+}
+
+
+TEST(fopenfclose, freopen_file)
+{
+	fd = fopen(stdpath, "w");
+	/* freopen() on opened file */
+	fd2 = freopen(stdpath, "w", fd);
```

Comment:
imho we need to asssert fd != NULL before doing freopen

Reply:
As `fd` is global pointer and it is not getting NULL-ed by fclose() the arbitrary `fd = NULL` was added in the `TEST_SETUP` so the condition is met. If test are badly written and do not close `fd` or `fd2` if opened then it may break tests. 

However this is also the case if local pointers are used, as then we are only sure that they are NULL in the beginning and if test does not properly `fclose()` the pointer then it is still undefined behavior (file may be still opened, thus blocked etc)

But after your mention I added asserts for pointers NULL-ability after opening and asserts for all `fclose()`-s

Reply:
fopen can fail and we do freaopen in the next line - it will crash in that case.